### PR TITLE
compute: Add Vector data type and functions

### PR DESCRIPTION
**Pull request by Lloyd-Pottiger**
### What problem does this PR solve?

Issue Number: ref #9032

Problem Summary:

### What is changed and how it works?

Pick https://github.com/tidbcloud/tiflash-cse/pull/153

```commit-message

```

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

**Review by breezewish — APPROVED** [review]

**Comment by breezewish** [comment]
Could also manually verify whether it works with these two PRs? Thank you!

- https://github.com/pingcap/tidb/pull/54635
- https://github.com/tikv/tikv/pull/17287

File: dbms/src/Flash/Coprocessor/TiDBColumn.h [hunk 1]
```diff
@@ -44,18 +44,21 @@ class TiDBColumn
     void encodeColumn(WriteBuffer & ss);
     void clear();
 
-private:
-    bool isFixed() { return fixed_size != VAR_SIZE; };
+    std::unique_ptr<WriteBufferFromOwnString> data;
     void finishAppendFixed();
     void finishAppendVar(UInt32 size);
```
Comment (JaySon-Huang) [inline dbms/src/Flash/Coprocessor/TiDBColumn.h:49]: must these method to be public?

File: dbms/src/Columns/ColumnNullable.cpp [hunk 2]
```diff
@@ -204,14 +204,29 @@ void ColumnNullable::get(size_t n, Field & res) const
...
 }
 
-StringRef ColumnNullable::getDataAt(size_t /*n*/) const
+StringRef ColumnNullable::getDataAt(size_t n) const
 {
-    throw Exception(fmt::format("Method getDataAt is not supported for {}", getName()), ErrorCodes::NOT_IMPLEMENTED);
+    if (!isNullAt(n))
+        return getNestedColumn().getDataAt(n);
+
+    throw Exception(
```
[the remark this answers is not in the document]
Comment (Lloyd-Pottiger) [reply dbms/src/Columns/ColumnNullable.cpp:212]: It is by design

File: dbms/src/TiDB/Decode/Vector.cpp [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+
+    for (size_t i = 0, i_max = size(); i < i_max; ++i)
+    {
+        // Hope this can be vectorized.
+        distance += elements[i] * b[i];
+        norma += elements[i] * elements[i];
+        normb += b[i] * b[i];
+    }
+
+    Float64 similarity
```
[the remark this answers is not in the document]
Comment (Lloyd-Pottiger) [reply dbms/src/TiDB/Decode/Vector.cpp:99]: In c++, `float / 0 = inf` and `0.0 / 0 = nan`

[the remark this answers is not in the document]
Comment (Lloyd-Pottiger) [reply dbms/src/TiDB/Decode/Vector.cpp:99]: If `norma` or `normb` equal to 0, `distance` must be 0.

[the remark this answers is not in the document]
Comment (Lloyd-Pottiger) [reply dbms/src/TiDB/Decode/Vector.cpp:99]: I will add some comments.

**Comment by wuhuizuo** [comment]
@Lloyd-Pottiger you can revert the commit: feb28b7d99bef70787a14b4f6ca8647f04c30440

I fixed it in https://github.com/PingCAP-QE/ci/pull/3042

**Review by JaySon-Huang — APPROVED** [review]

**Pull request merged**